crates/bevy_ecs/src/query/state.rs

```diff
@@ -526,12 +561,13 @@ impl<Q: WorldQuery, F: WorldQuery> QueryState<Q, F> {
     pub fn iter_combinations<'w, 's, const K: usize>(
         &'s mut self,
         world: &'w World,
-    ) -> QueryCombinationIter<'w, 's, Q, F, K> {
```

Comment:
this is a soundness fix, but it would be possible to make a much smaller PR solely fixing this soundness bug if we wanted (I would rather just do this PR though)